### PR TITLE
openshift-appliance: added e2e jobs (ipv4 and dualstack)

### PR DESCRIPTION
**Pull request by danielerez**
Added the following e2e jobs to openshift-appliance:
* e2e-compact-ipv4-static
* e2e-sno-dualstack-dhcp

### CODE REVIEW

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-e2e-compact-ipv4-static

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-e2e-compact-ipv4-static

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-e2e-compact-ipv4-static

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-e2e-compact-ipv4-static

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-e2e-compact-ipv4-static

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-e2e-compact-ipv4-static

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-e2e-compact-ipv4-static

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-e2e-compact-ipv4-static

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-e2e-sno-ipv4-dhcp

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-e2e-compact-ipv4-static

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-e2e-compact-ipv4-static

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-e2e-ha-dualstack-dhcp

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-e2e-sno-ipv4-dhcp

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-e2e-sno-ipv4-dhcp

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-e2e-compact-ipv4-static

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-e2e-compact-ipv4-static

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-e2e-sno-dualstack-dhcp

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-installer-master-e2e-agent-compact-ipv4

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-e2e-sno-dualstack-dhcp

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-e2e-compact-ipv4-static

**Review by osherdp — APPROVED** [review]

**Comment by osherdp** [comment]
/retitle openshift-appliance: added e2e jobs (ipv4 and dualstack)

File: ci-operator/step-registry/baremetalds/devscripts/setup/baremetalds-devscripts-setup-ref.yaml [hunk 1]
```diff
@@ -25,5 +25,7 @@ ref:
     env: OPENSHIFT_INSTALL_RELEASE_IMAGE
   - name: "release:latest"
     env: OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE
+  - name: "agent-preinstall-image-builder"
+    env: APPLIANCE_IMAGE
```
Comment (andfasano) [inline ci-operator/step-registry/baremetalds/devscripts/setup/baremetalds-devscripts-setup-ref.yaml:29]: I'm not sure about this one, couldn't be simply set via DEVSCRIPTS_CONFIG? The step is shared by many different jobs (from different teams), and it should have only very generic env vars

Comment (danielerez) [reply ci-operator/step-registry/baremetalds/devscripts/setup/baremetalds-devscripts-setup-ref.yaml:29]: It should be done using dependencies as it translates the url (to the local image that is built in job). Couldn't find a more isolated way for it...

Comment (andfasano) [reply ci-operator/step-registry/baremetalds/devscripts/setup/baremetalds-devscripts-setup-ref.yaml:29]: As for many other workflows based on this step specific dev-scripts configuration must be placed inside DEVSCRIPTS_CONFIG, that's the mechanism to keep the step simple and generic. The step offers also an additional customization mechanism (dev-scripts-additional-config) for very specific needs, but the preferred way (also for the maintenance) remains DEVSCRIPTS_CONFIG

Comment (danielerez) [reply ci-operator/step-registry/baremetalds/devscripts/setup/baremetalds-devscripts-setup-ref.yaml:29]: @andfasano So the issue is with translating the image url, not the actual mechanism for adding it to dev-scripts.
I.e. see the setup ob [log](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/42043/rehearse-42043-pull-ci-openshift-appliance-master-e2e-sno-dualstack-dhcp/1688832932581478400/artifacts/e2e-sno-dualstack-dhcp/baremetalds-devscripts-setup/build-log.txt):
```
export APPLIANCE_IMAGE=registry.build05.ci.openshift.org/ci-op-hn18sxx2/pipeline@sha256:7a4c03459d76ee47f7ec0657abe1648855d6ddbe28673563d0376606c192d5d6
```
I.e. the value can't be hard-coded. Any suggestion for adding it without using dependencies?

Comment (osherdp) [reply ci-operator/step-registry/baremetalds/devscripts/setup/baremetalds-devscripts-setup-ref.yaml:29]: I don't think there's a way to do that without the dependencies mechanism (which is the main method in prow to pass built/imported images)

Comment (danielerez) [reply ci-operator/step-registry/baremetalds/devscripts/setup/baremetalds-devscripts-setup-ref.yaml:29]: After discussing with @andfasano, added a condition to 'baremetalds-devscripts-setup-commands.sh' for ensuring that the APPLIANCE_IMAGE var is added only for appliance e2e jobs.

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-e2e-compact-ipv4-static

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-e2e-sno-dualstack-dhcp

File: ci-operator/step-registry/baremetalds/devscripts/setup/baremetalds-devscripts-setup-ref.yaml (hunk 1 above)
Comment (andfasano) [reply ci-operator/step-registry/baremetalds/devscripts/setup/baremetalds-devscripts-setup-ref.yaml:29]: Unfortunately the dependency definition it's possible only at the step level (while the override can be set elsewhere) and yes, there's no other way currently available to inject a pullspec. The current approach at least avoids setting the APPLIANCE_IMAGE var when not required

**Comment by andfasano** [comment]
Let's run also a standard metal job, to verify that the changes do not impact them:

/pj-rehearse pull-ci-openshift-kubernetes-release-4.14-e2e-metal-ipi

**Comment by danielerez** [comment]
> Let's run also a standard metal job, to verify that the changes do not impact them:
> 
> /pj-rehearse pull-ci-openshift-kubernetes-release-4.14-e2e-metal-ipi

Seems like it didn't like the 'none' image:
```
* could not run steps: step e2e-metal-ipi failed: "e2e-metal-ipi" pre steps failed: could not determine image pull spec for image none on step baremetalds-devscripts-setup
```

**Comment by andfasano** [comment]
> > Let's run also a standard metal job, to verify that the changes do not impact them:
> > /pj-rehearse pull-ci-openshift-kubernetes-release-4.14-e2e-metal-ipi
> 
> Seems like it didn't like the 'none' image:
> 
> ```
> * could not run steps: step e2e-metal-ipi failed: "e2e-metal-ipi" pre steps failed: could not determine image pull spec for image none on step baremetalds-devscripts-setup
> ```

If the dependency must be _always_ defined and cannot be empty, I guess at this point the only option left is to constrain the APPLIANCE_IMAGE injection to the presence of another config var, ie `AGENT_E2E_TEST_BOOT_MODE == DISKIMAGE`

**Comment by danielerez** [comment]
> > > Let's run also a standard metal job, to verify that the changes do not impact them:
> > > /pj-rehearse pull-ci-openshift-kubernetes-release-4.14-e2e-metal-ipi
> > 
> > 
> > Seems like it didn't like the 'none' image:
> > ```
> > * could not run steps: step e2e-metal-ipi failed: "e2e-metal-ipi" pre steps failed: could not determine image pull spec for image none on step baremetalds-devscripts-setup
> > ```
> 
> If the dependency must be _always_ defined and cannot be empty, I guess at this point the only option left is to constrain the APPLIANCE_IMAGE injection to the presence of another config var, ie `AGENT_E2E_TEST_BOOT_MODE == DISKIMAGE`

Changed, trying again..

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-kubernetes-release-4.14-e2e-metal-ipi

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-e2e-sno-dualstack-dhcp

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-e2e-compact-ipv4-static

**Comment by andfasano** [comment]
@danielerez as a last check, would you mind trying also the `APPLIANCE_IMAGE=${APPLIANCE_IMAGE}` in the `DEVSCRIPTS_CONFIG` var? If it works then it wouldn't be necessary anymore the [change](https://github.com/openshift/release/pull/42043/files#diff-f891b30b93f62c85827a6dbf087d3deb2ce2f35e6c73cafeed7e4787d1b7b386R205-R210) in the devscripts-setup script

**Comment by danielerez** [comment]
> @danielerez as a last check, would you mind trying also the `APPLIANCE_IMAGE=${APPLIANCE_IMAGE}` in the `DEVSCRIPTS_CONFIG` var? If it works then it wouldn't be necessary anymore the [change](https://github.com/openshift/release/pull/42043/files#diff-f891b30b93f62c85827a6dbf087d3deb2ce2f35e6c73cafeed7e4787d1b7b386R205-R210) in the devscripts-setup script

Yeah, tried it already. It doesn't work:(

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-kubernetes-release-4.14-e2e-metal-ipi

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-e2e-compact-ipv4-static

**Comment by danielerez** [comment]
> > @danielerez as a last check, would you mind trying also the `APPLIANCE_IMAGE=${APPLIANCE_IMAGE}` in the `DEVSCRIPTS_CONFIG` var? If it works then it wouldn't be necessary anymore the [change](https://github.com/openshift/release/pull/42043/files#diff-f891b30b93f62c85827a6dbf087d3deb2ce2f35e6c73cafeed7e4787d1b7b386R205-R210) in the devscripts-setup script
> 
> Yeah, tried it already. It doesn't work:(

That's the job [log](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/42007/rehearse-42007-pull-ci-openshift-appliance-master-e2e-sno-dualstack-dhcp/1689927801194418176/artifacts/e2e-sno-dualstack-dhcp/baremetalds-devscripts-setup/build-log.txt). It uses the default value instead of the one on pipeline:
```
++(/root/dev-scripts/agent/common.sh:15): source(): export APPLIANCE_IMAGE=quay.io/edge-infrastructure/openshift-appliance:latest

```

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-appliance-master-e2e-sno-dualstack-dhcp

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-kubernetes-release-4.14-e2e-metal-ipi

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-cluster-api-provider-baremetal-release-4.7-e2e-metal-ipi

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-cluster-api-provider-baremetal-release-4.14-e2e-metal-ipi

**Comment by danielerez** [comment]
@andfasano @osherdp Some 4.7 jobs seem to fail due to missing appliance image:
```
  * could not run steps: step e2e-metal-ipi failed: "e2e-metal-ipi" pre steps failed: could not determine image pull spec for image agent-preinstall-image-builder on step baremetalds-devscripts-setup
```
E.g. cluster-api-provider-baremetal/release-4.7/e2e-metal-ipi [job](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/42043/rehearse-42043-pull-ci-openshift-cluster-api-provider-baremetal-release-4.7-e2e-metal-ipi/1690630242361675776). Though release-4.14/e2e-metal-ipi [job](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/42043/rehearse-42043-pull-ci-openshift-cluster-api-provider-baremetal-release-4.14-e2e-metal-ipi/1690630334091104256) looks fine.
Any idea how to mitigate this issue?

**Comment by danielerez** [comment]
> **assisted-service-master-edge-e2e-ai-operator-ztp-3masters-periodic ** — Job failed.                     BaseSHA:98e90a4acae6989b281d9c564eb6f414a9581ef2

Same issue also in assisted-service-master-edge-e2e-ai-operator-ztp-3masters-periodic job: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_release/42043/rehearse-42043-periodic-ci-openshift-assisted-service-master-edge-e2e-ai-operator-ztp-3masters-periodic/1690411493440360448/build-log.txt
```
  * could not run steps: step e2e-ai-operator-ztp-3masters-periodic failed: "e2e-ai-operator-ztp-3masters-periodic" pre steps failed: could not determine image pull spec for image agent-preinstall-image-builder on step baremetalds-devscripts-setup 
```

**Comment by osherdp** [comment]
> agent-preinstall-image-builder

it probably means that ``agent-preinstall-image-builder`` needs to be imported in every place that uses ``baremetalds-devscripts-setup`` indirectly

**Comment by osherdp** [comment]
> > agent-preinstall-image-builder
> 
> it probably means that `agent-preinstall-image-builder` needs to be imported in every place that uses `baremetalds-devscripts-setup` indirectly

https://redhat-internal.slack.com/archives/CBN38N3MW/p1640707502146600

**Comment by danielerez** [comment]
> /pj-rehearse pull-ci-openshift-cluster-api-provider-baremetal-release-4.7-e2e-metal-ipi

ok, did the same trick then. I.e. use an existing image for on baremetalds-devscripts-setup-ref.yaml, and override with the actual image on relevant jobs.

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-cluster-api-provider-baremetal-release-4.7-e2e-metal-ipi

**Comment by danielerez** [comment]
/pj-rehearse ci/rehearse/openshift/installer/master/e2e-agent-compact-ipv4-appliance

**Comment by danielerez** [comment]
/pj-rehearse pull-ci-openshift-installer-master-e2e-agent-compact-ipv4-appliance

**Comment by danielerez** [comment]
> > /pj-rehearse pull-ci-openshift-cluster-api-provider-baremetal-release-4.7-e2e-metal-ipi
> 
> ok, did the same trick then. I.e. use an existing image for on baremetalds-devscripts-setup-ref.yaml, and override with the actual image on relevant jobs.

@osherdp @andfasano Rehearsals look ok now, can you please take another look?

**Comment by danielerez** [comment]
@derekhiggins can you please take a look at the 'baremetalds-devscripts-setup' step?

**Pull request merged**